### ISSUE TITLE
Switch to building manylinux wheels?

### ISSUE THREAD

**Issue by matthew-brett** [matthew-brett/travis-wheel-builder#23]
It seems like several projects are using these wheels, but they are not going to work for other Ubuntus.

Recently, the `manylinux` proposal got accepted by the Python packaging authority - see refs in https://github.com/pypa/manylinux

This defines a wheel build format that will work across many Linux variants, but particularly, any recent Ubuntu, as for the travis configurations.

These wheels are a little harder to build than we have here, and we'd have to use [docker](https://docs.travis-ci.com/user/docker/) - but it wouldn't be enormously hard to build up a range of wheels that could be used in almost any Linux build / test environment.

I already drafted up some wheel building scripts here: https://github.com/matthew-brett/manylinux-builds

I have a preliminary set of wheels here: http://nipy.bic.berkeley.edu/manylinux/

Any interest in getting that going?

**Comment by blink1073** [comment]
How did you build the existing wheels?

**Comment by matthew-brett** [comment]
I built them on a beefy machine on campus using the command at the top of the individual build files - e.g. https://github.com/matthew-brett/manylinux-builds/blob/master/build_numpies.sh#L3

**Comment by blink1073** [comment]
Okay, so we'd target a Travis build similar to what we did for this repo.  I have exactly zero experience with docker, however.

**Comment by matthew-brett** [comment]
I highly recommend it - it's got good documentation, and it's often as simple as finding the right image (which is the manylinux image in the command), and starting it up to get the OS you want, working as root.   It's quick and vastly better than VMs.

**Comment by blink1073** [comment]
Ideally individual repos would be able to build their own wheels using that container and release them on PyPI, but I can certainly see a use for this, especially when older  versions of libs are desired.

**Comment by matthew-brett** [comment]
Yes, sure, that's the hope, that projects will build their own, and eventually people can get everything they need from pip.   I guess that will take a little while though - at least six months.  Then, as you say, there's the old versions of things to test against.  So I think the idea of this repo is still useful.

**Comment by blink1073** [comment]
Ha, crap, I tried to build hdf5 and got this:

```The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).

The job has been terminated```

https://travis-ci.org/blink1073/manylinux-builds/builds/115516242

**Comment by blink1073** [comment]
But if you don't pipe to stdout, the build may time out.  My current attempt is to break up the commands, print the command, but pipe the longer output to `/dev/null`.

**Comment by blink1073** [comment]
Progress, but I'm missing something:

```
tar: Error is not recoverable: exiting now
The command "docker run --rm -v $PWD:/io quay.io/pypa/manylinux1_x86_64 /io/build_hdf5s.sh" exited with 141.
0.01s$ ls libraries
ls: cannot access libraries: No such file or directory
The command "ls libraries" exited with 2.
Done. Your build exited with 1.
```

https://travis-ci.org/blink1073/manylinux-builds/builds/115520563

**Comment by matthew-brett** [comment]
Yes, sorry, my fault, I was assuming that a `libraries` directory already existed.

I've refactored the scripts a bit, now trying to work out how to prevent script bombing when doesn't find all tar members.

**Comment by matthew-brett** [comment]
Nope, still futzing...

**Comment by matthew-brett** [comment]
OK - I think the `build_hdf5s.sh` script works now.

**Comment by matthew-brett** [comment]
By the way, you probably saw, but you can set environment variables - like `HDF5_VERSIONS` - with something like:

```
docker run -e HDF5_VERSIONS=1.8.3 --rm -v $PWD:/io quay.io/pypa/manylinux1_x86_64 /io/build_hdf5s.sh
```

**Comment by blink1073** [comment]
Success!

```bash
The command "docker run -e HDF5_VERSIONS=1.8.3 --rm -v $PWD:/io quay.io/pypa/manylinux1_x86_64 /io/build_hdf5s.sh" exited with 0.
0.01s$ ls libraries
hdf5-1.8.3.tgz
The command "ls libraries" exited with 0.
```

**Comment by matthew-brett** [comment]
Nice!

**Comment by matthew-brett** [comment]
Following up on this one.

I've got a good stash of wheels currently hosted on a Berkeley server here : https://nipy.bic.berkeley.edu/manylinux .  I'm pretty confident that these wheels are OK.

That server is up to date, and currently in good shape, but it would be good to switch to something we can rely on in the longer term.

So, the question is should we:

* Copy the manylinux wheels into travis-wheels.scikit-image.org - and just build up those wheels in the old container, or:
* Start a new container, just for the manylinux wheels.

It's possible that, copying the new wheels into the old container would break something for someone - although I don't yet know how that would happen.    On the other hand, it does give a smooth upgrade path for anyone already using these, meaning that, for example, they may be able to switch to trusty builds with no change in their setup.

What do you think @blink1073 ?

**Comment by blink1073** [comment]
I agree that it would be cleaner to create a `manylinux-wheels.scikit-image.org` and start fresh there, for both backward compatibility and cleanliness.

**Comment by stefanv** [comment]
Let me know if you need DNS set up.
​
